File: src/game/activity/xp.ts
import type { GameDatabase } from '../data/types'
import type { PlayerSave, SkillProgress } from '../save/types'

export function getSkillProgress(save: PlayerSave, skillId: string): SkillProgress {
  const existing = save.skills.find((skill) => skill.skillId === skillId)
  if (existing) return existing
  return { skillId, level: 1, xp: 0 }
}

export function levelForTotalXp(db: GameDatabase, totalXp: number): number {
  let level = 1
  for (const row of db.XPCurve) {
    if (totalXp >= row['Total XP at Level']) {
      level = row.Level
    } else {
      break
    }
  }
  return level
}

export function applyXp(
  save: PlayerSave,
  db: GameDatabase,
  skillId: string,
  xpAmount: number,
): { save: PlayerSave; leveledUpTo: number | null } {
  if (xpAmount <= 0) return { save, leveledUpTo: null }

  const skills = save.skills.map((skill) => ({ ...skill }))
  let progress = skills.find((skill) => skill.skillId === skillId)
  if (!progress) {
    progress = { skillId, level: 1, xp: 0 }
    skills.push(progress)
  }

  const previousLevel = progress.level
  progress.xp += xpAmount
  progress.level = levelForTotalXp(db, progress.xp)
  const leveledUpTo = progress.level > previousLevel ? progress.level : null

  return {
    save: { ...save, skills },
    leveledUpTo,
  }
}

/** Raise a skill to at least `minLevel` (by total XP). No-op if already there or higher. */
export function raiseSkillToMinimumLevel(
  save: PlayerSave,
  db: GameDatabase,
  skillId: string,
  minLevel: number,
): { save: PlayerSave; raised: boolean } {
  if (minLevel <= 1) return { save, raised: false }
  const current = getSkillProgress(save, skillId)
  if (current.level >= minLevel) return { save, raised: false }

  const xpAtLevel = db.XPCurve.find((row) => row.Level === minLevel)?.['Total XP at Level']
  if (typeof xpAtLevel !== 'number') return { save, raised: false }

  const skills = save.skills.map((skill) => ({ ...skill }))
  let progress = skills.find((skill) => skill.skillId === skillId)
  if (!progress) {
    progress = { skillId, level: 1, xp: 0 }
    skills.push(progress)
  }
  progress.xp = Math.max(progress.xp, xpAtLevel)
  progress.level = levelForTotalXp(db, progress.xp)

  return {
    save: { ...save, skills },
    raised: progress.level >= minLevel && current.level < minLevel,
  }
}

File: packages/ik_rules/lib/src/activity/xp.dart
import 'dart:math' as math;

import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../save/generated/save_models.dart';

SkillProgress getSkillProgress(PlayerSave save, String skillId) {
  final existing = save.skills.firstWhereOrNull((skill) => skill.skillId == skillId);
  return existing ?? SkillProgress(skillId: skillId, level: 1, xp: 0);
}

num levelForTotalXp(GameDatabase db, num totalXp) {
  num level = 1;
  for (final row in db.xpCurve) {
    if (totalXp >= row.totalXpAtLevel) {
      level = row.level;
    } else {
      break;
    }
  }
  return level;
}

class ApplyXpResult {
  const ApplyXpResult({required this.save, this.leveledUpTo});

  final PlayerSave save;
  final num? leveledUpTo;
}

ApplyXpResult applyXp(PlayerSave save, GameDatabase db, String skillId, num xpAmount) {
  if (xpAmount <= 0) return ApplyXpResult(save: save);

  final skills = [...save.skills];
  var index = skills.indexWhere((skill) => skill.skillId == skillId);
  if (index < 0) {
    skills.add(SkillProgress(skillId: skillId, level: 1, xp: 0));
    index = skills.length - 1;
  }

  final previous = skills[index];
  final xp = previous.xp + xpAmount;
  final level = levelForTotalXp(db, xp);
  skills[index] = previous.copyWith(xp: xp, level: level);

  return ApplyXpResult(
    save: save.copyWith(skills: skills),
    leveledUpTo: level > previous.level ? level : null,
  );
}

class RaiseSkillResult {
  const RaiseSkillResult({required this.save, required this.raised});

  final PlayerSave save;
  final bool raised;
}

/// Raises a skill to at least [minLevel] by total XP; a no-op when already there.
RaiseSkillResult raiseSkillToMinimumLevel(
  PlayerSave save,
  GameDatabase db,
  String skillId,
  num minLevel,
) {
  if (minLevel <= 1) return RaiseSkillResult(save: save, raised: false);
  final current = getSkillProgress(save, skillId);
  if (current.level >= minLevel) return RaiseSkillResult(save: save, raised: false);

  final curveRow = db.xpCurve.firstWhereOrNull((row) => row.level == minLevel);
  if (curveRow == null) return RaiseSkillResult(save: save, raised: false);
  final xpAtLevel = curveRow.totalXpAtLevel;

  final skills = [...save.skills];
  var index = skills.indexWhere((skill) => skill.skillId == skillId);
  if (index < 0) {
    skills.add(SkillProgress(skillId: skillId, level: 1, xp: 0));
    index = skills.length - 1;
  }

  final xp = math.max(skills[index].xp, xpAtLevel);
  final level = levelForTotalXp(db, xp);
  skills[index] = skills[index].copyWith(xp: xp, level: level);

  return RaiseSkillResult(
    save: save.copyWith(skills: skills),
    raised: level >= minLevel && current.level < minLevel,
  );
}
